devlxd: Volume management API

MusicDin opened the pull request:
This PR adds the devLXD endpoints that allow volume management, primarily meant for LXD CSI driver.
This includes volume creation, deletion, and modification, as well as instance device management, which allows attaching and detaching volumes.

### Access

The endpoints are disabled by default and require `security.devlxd.management.volumes` to be enabled on a particular instance. Additionally, endpoints `/1.0/storage-pools/*` and `/1.0/instances/*` require a caller to authenticate using a valid devLXD bearer token. Finally, devLXD allows attaching and detaching devices from other instances within the same project, if the identity has sufficient permissions. 

### Ownership

DevLXD allows only management of owned volumes and instance devices. This means that only volumes created by the devLXD identity can be later managed or removed.

Volume owner is stored in volume configuration `volatile.owner`, while instance device ownership is stored in instance configuration `volatile.<device>.devlxd.owner`. In both cases, the value is set to devLXD identity identifier. 

---

Rebase after:
- [x] https://github.com/canonical/lxd/pull/16329
- [x] https://github.com/canonical/lxd/pull/16450
- [x] https://github.com/canonical/lxd/pull/16451

tomponline commented on lxd/response/response.go:
```diff
@@ -65,6 +81,10 @@ func (r *devLXDResponse) Render(w http.ResponseWriter, req *http.Request) (err e
 		}
 
+		if etag != "" {
+			w.Header().Set("ETag", `"`+etag+`"`)
```
Needs a comment.

tomponline commented on lxd/devlxd.go:
```diff
@@ -650,6 +652,26 @@ func allowDevLXDPermission(entityType entity.Type, entitlement auth.Entitlement,
 
+// getDevLXDIdentity extracts the identity of the caller from the request context.
+// The identity is expected to be present and have a valid identifier. Otherwise, an error is returned.
+func getDevLXDIdentity(ctx context.Context) (*identity.CacheEntry, error) {
```
Do we need a devlxd specific one of these, it seems to not have anything devlxd specific in it?

MusicDin replied:
We can make it generic (`getIdentity()`?). The main intention of the function is just to avoid duplication.

tomponline commented on lxd/devlxd.go:
```diff
@@ -556,7 +556,19 @@ func registerDevLXDEndpoint(d *Daemon, apiRouter *mux.Router, apiVersion string,
+			}
+
+			// If the request is not trusted, only call the handler
+			// if the action allows it.
```
No need to wrap the comment

tomponline commented on lxd/devlxd_util.go:
```diff
@@ -63,3 +68,101 @@ func NewRequestWithContext(ctx context.Context, method string, url string, data
+}
+
+// NewResponseCapture creates a new responseCapture for the given request.
+func NewResponseCapture(req *http.Request) *responseCapture {
```
Can these go into the `response` package, they dont appear to be devlxd specific?

tomponline commented on lxd/devlxd.go:
```diff
@@ -146,11 +146,39 @@ func devLXDAPIGetHandler(d *Daemon, r *http.Request) response.Response {
+
+		info := &api.Server{}
+
+		resp := api10Get(d, req)
```
Rather than calling `api10Get` here (which gets a lot of other info unnecessarily here) and then discard all but the storage drivers info, please can you factor out the supported storage driver code from `api10Get` into its own function which you can call from `api10Get` and here.

tomponline commented on shared/api/devlxd_storage.go:
```diff
@@ -0,0 +1,30 @@
+
+	// Storage pool configuration map (refer to doc/storage.md)
+	// Example: {"volume.block.filesystem": "ext4", "volume.size": "50GiB"}
+	Config map[string]string `json:"config" yaml:"config"`
```
@MusicDin @markylaing we need to be very careful here as storage pool config can contain passwords to the external storage appliances. 

IIRC we only grant access to storage pool config when you are an admin, is that right @markylaing ?

@MusicDin whats the reason for needing access to this Config map?

Perhaps we need to move the pool level secret keys into the new `secrets` table and relax access to Pool config first before this change lands.

tomponline commented on shared/api/devlxd_storage.go:
```diff
@@ -13,18 +13,73 @@ type DevLXDStoragePool struct {
+	Type string `json:"type" yaml:"type"`
+
+	// Migration source.
+	Source StorageVolumeSource `json:"source" yaml:"source"`
```
Why do we need this?

tomponline replied:
Yes if there's additional PRs coming lets do that - they can all some under the single API extension though

tomponline commented on lxd/devlxd_storage.go:
```diff
@@ -46,3 +47,83 @@ func devLXDStoragePoolGetHandler(d *Daemon, r *http.Request) response.Response {
+
+	respVols := make([]api.DevLXDStorageVolume, 0, len(vols))
+	for _, vol := range vols {
+		if vol.Config["volatile.owner"] != identity.Identifier {
```
This is easy to miss, lets break it out into a separate function and then add a comment explaining what is happening here.

tomponline commented on lxd/devlxd_storage.go:
```diff
@@ -127,3 +130,75 @@ func devLXDStoragePoolVolumesGetHandler(d *Daemon, r *http.Request) response.Res
 	return response.DevLXDResponse(http.StatusOK, respVols, "json")
 }
+
+func devLXDStoragePoolVolumesPostHandler(d *Daemon, r *http.Request) response.Response {
```
missing doc block

tomponline commented on lxd/devlxd_storage.go:
```diff
@@ -46,3 +47,83 @@ func devLXDStoragePoolGetHandler(d *Daemon, r *http.Request) response.Response {
+	Get:  devLXDAPIEndpointAction{Handler: devLXDStoragePoolVolumesGetHandler, AccessHandler: allowDevLXDAuthenticated},
+}
+
+func devLXDStoragePoolVolumesGetHandler(d *Daemon, r *http.Request) response.Response {
```
missing doc block

tomponline commented on lxd/devlxd_storage.go:
```diff
@@ -127,3 +130,75 @@ func devLXDStoragePoolVolumesGetHandler(d *Daemon, r *http.Request) response.Res
+		return response.DevLXDErrorResponse(api.NewStatusError(http.StatusBadRequest, "URL volume type does not match the volume type in body"))
+	}
+
+	// Set caller's identity ID as the volume owner.
```
Please expand the comment to explain why.

tomponline replied:
this point as not been addressed yet

tomponline commented on lxd/devlxd_storage.go:
```diff
@@ -202,3 +203,80 @@ func devLXDStoragePoolVolumesPostHandler(d *Daemon, r *http.Request) response.Re
+	return respVol, etag, nil
+}
+
+func devLXDStoragePoolVolumeGetHandler(d *Daemon, r *http.Request) response.Response {
```
missing doc block

tomponline commented on lxd/devlxd_storage.go:
```diff
@@ -280,3 +281,46 @@ func devLXDStoragePoolVolumeGetHandler(d *Daemon, r *http.Request) response.Resp
 	return response.DevLXDResponseETag(http.StatusOK, vol, "json", etag)
 }
+
+func devLXDStoragePoolVolumeDeleteHandler(d *Daemon, r *http.Request) response.Response {
```
missing doc block

tomponline commented on lxd/devlxd_storage.go:
```diff
@@ -282,6 +283,83 @@ func devLXDStoragePoolVolumeGetHandler(d *Daemon, r *http.Request) response.Resp
 	return response.DevLXDResponseETag(http.StatusOK, vol, "json", etag)
 }
 
+func devLXDStoragePoolVolumePutHandler(d *Daemon, r *http.Request) response.Response {
```
missing doc block

tomponline commented on lxd/storage/utils.go:
```diff
@@ -545,6 +545,15 @@ func poolAndVolumeCommonRules(vol *drivers.Volume) map[string]func(string) error
+		//  defaultdesc: owner's identity ID
+		//  shortdesc: The identity ID of the volume owner
+		//  scope: global
+		rules["volatile.owner"] = validate.Optional(validate.IsUUID)
```
looking at the previous commit, for the device level `devlxd.owner` key, should this be also `volatile.devlxd.owner` for consistency?

tomponline commented on lxd/devlxd_instances.go:
```diff
@@ -0,0 +1,21 @@
+	diskDeviceAllowed := hasInstanceSecurityFeatures(inst, devLXDSecurityMgmtVolumesKey)
+
+	return func(device map[string]string) bool {
+		return filters.IsCustomVolumeDisk(device) && diskDeviceAllowed
```
Suggest switching round the conditions here as its cheaper to check a boolean first.

tomponline commented on lxd/devlxd_instances.go:
```diff
@@ -3,10 +3,29 @@ package main
+// getOwnedDevices extracts instance devices that are owned by the provided identity.
+// Two maps are returned, one containing owned devices, and the other containing the remaining
+// devices.
+func getOwnedDevices(inst api.Instance, identityID string) (ownedDevices map[string]map[string]string, otherDevices map[string]map[string]string) {
```
remember you are in the `main` package and so IDEs will suggest functions to use for the whole of LXD.

So make sure that if a function or variable in the main package is devlxd specific (as this one is) that its prefixed with `devLXD`.

tomponline commented on lxd/devlxd_instances.go:
```diff
@@ -3,10 +3,29 @@ package main
 
+// getOwnedDevices extracts instance devices that are owned by the provided identity.
+// Two maps are returned, one containing owned devices, and the other containing the remaining
+// devices.
```
no need to wrap here.

tomponline commented on lxd/devlxd_instances.go:
Seeing the amount of devlxd specific functionality going into the main package I am starting to think its time to split devlxd specific functionality into its own `devlxd` package.

tomponline commented:
I'm thinking for future readers' benefit it would be helpful to expand the commit message for each of the API endpoints you are adding explaining why they are being added and what purpose they serve for the CSI. Rather than just a one line naming the endpoint as we have now. Thanks

tomponline commented on lxd/devlxd_instances.go:
```diff
@@ -4,13 +4,74 @@ import (
+	Get:  devLXDAPIEndpointAction{Handler: devLXDInstanceGetHandler, AccessHandler: allowDevLXDPermission(entity.TypeInstance, auth.EntitlementCanView, "name")},
+}
+
+func devLXDInstanceGetHandler(d *Daemon, r *http.Request) response.Response {
```
missing doc block

tomponline commented on lxd/devlxd_instances.go:
```diff
@@ -4,13 +4,74 @@ import (
+	}
+
+	// Get owned devices.
+	ownedDevices, _ := getOwnedDevices(targetInst, identity.Identifier)
```
does getOwnedDevices need to return other devices at all?

MusicDin replied:
No. Previously we used PUT and we had to retain non-owned devices in the request. Now that we use patch, this is no longer needed.

tomponline commented on lxd/devlxd_instances.go:
```diff
@@ -4,13 +4,74 @@ import (
+		}
+	}
+
+	// Map to devLXD type.
```
Suggest "Convert" to avoid confusion with built-in type `map`

tomponline commented on lxd/devlxd_storage.go:
```diff
@@ -349,7 +349,7 @@ func devLXDStoragePoolVolumePutHandler(d *Daemon, r *http.Request) response.Resp
 	err = addStoragePoolVolumeDetailsToRequestContext(d.State(), req)
 	if err != nil {
-		response.DevLXDErrorResponse(err)
+		return response.DevLXDErrorResponse(err)
```
this should be added to the original commit that added `devLXDStoragePoolVolumePutHandler`

tomponline commented on lxd-agent/response.go:
```diff
@@ -12,19 +13,38 @@ type devLXDResponse struct {
+
+	// Set ETag if provided.
+	if r.etag != "" {
+		// Remove quotes from ETag if present.
```
Still not clear on this quoting business. Seems like we should be breaking this out into separate function(s) to de-dupe this quoting logic as ive seen it in a few places now.

tomponline commented on lxd-agent/response.go:
```diff
@@ -65,6 +85,15 @@ func okResponse(ct any, ctype string) *devLXDResponse {
 	}
 }
 
+func okResponseETag(ct any, ctype string, etag string) *devLXDResponse {
```
This got me thinking, why does the main LXD not need an `okResponseEtag`, but devlxd does?

MusicDin replied:
LXD has `SyncResponseEtag()`

tomponline commented on lxd/instance/drivers/driver_qemu.go:
```diff
@@ -5771,6 +5771,7 @@ func (d *qemu) Update(args db.InstanceArgs, userRequested bool) error {
 			"security.csm",
 			"security.devlxd",
 			"security.devlxd.images",
+			"security.devlxd.management.volumes",
```
do we need the same thing for containers?

tomponline reviewed (changes requested):
Thanks @MusicDin looking good so far.

I've been trying to think of a way we can split this up into smaller PRs as its rather unwieldy as it stands.

How about we break this PR up into the following 2 PRs:

1. LXD side devlxd APIs, devlxd client and associated container tests.
2. lxd-agent side devlxd APIs to add VM support.

They can both be under the same API extension though.

The only big area of concern I saw was the storage pool Config field.
I did not see any mechanism to prevent leaking storage appliance credentials to devlxd users.

MusicDin commented on lxd/response/response.go:
```diff
@@ -51,12 +52,27 @@ type Response interface {
 func (r *devLXDResponse) Render(w http.ResponseWriter, req *http.Request) (err error) {
+	etag := r.etag
+
+	// Strip quotes from ETag if present, as the Etag may be quoted if received
```
Because LXD quotes ETag in sync response:
https://github.com/canonical/lxd/blob/d0301013b227c5915dbe2330431e0795d4ad7507/lxd/response/response.go#L176-L184

MusicDin commented on lxd/response/response.go:
```diff
@@ -74,6 +94,10 @@ func (r *devLXDResponse) Render(w http.ResponseWriter, req *http.Request) (err e
 	}
 
+	if etag != "" {
+		w.Header().Set("ETag", `"`+etag+`"`)
```
It's always quoted and in case of devLXD it ensures it is first unquoted to prevent double quoting.

MusicDin commented on lxd/devlxd_instances.go:
```diff
@@ -1,13 +1,95 @@
+// - Remove:
+//   - Condition: Existing device that devLXD can manage is set to "null" in the request.
+//   - Action:    Removes existing device from the instance and removes its owner from the instance config.
+func patchInstanceDevices(inst *api.Instance, req api.DevLXDInstancePut, identityID string, isDeviceAccessible devLXDDeviceAccessValidator) error {
```
Done.

MusicDin commented:
> I've been trying to think of a way we can split this up into smaller PRs as its rather unwieldy as it stands.
> How about we break this PR up into the following 2 PRs:
> - LXD side devlxd APIs, devlxd client and associated container tests.
> - lxd-agent side devlxd APIs to add VM support.

I've separated the PR as per your suggestion. Also, removed everything snapshots and volume cloning related, which was still in the PR (for example: volume `Restore`/`Source` fields) as this will also land in separate PRs.

markylaing commented on doc/.wordlist.txt:
```diff
@@ -7,6 +7,7 @@ APIs
 balancer
 Charmhub
 CLI
+devLXD
```
Shouldn't need this. I think we should be using DevLXD everywhere (or `devlxd` if necessary).

markylaing commented on lxd/api_vsock.go:
```diff
@@ -37,6 +37,14 @@ func hoistReqVM(d *Daemon, r *http.Request, handler devLXDAPIHandlerFunc) respon
+			return resp
+		}
+	}
+
```
I'm a little confused as to why the access handler needs to go in the hoist functions.

The hoist functions were used to determine the `instance.Instance` before passing it into the handlers, which at the time, took an `instance.Instance` as an argument. 

Since we now store the instance in the request context, we no longer need to have the hoist functions work this way. 

I think we should have a separate PR which makes this change and introduces access handlers to the devlxd API.

markylaing replied:
> We could move the call to the handler from hoist func. Then this would allow us to first call hoist func to retrieve the instance, and later call the remaining handlers.

Yeah that's what I'm suggesting. I think these can now be functions that return the instance or an error, and are called at the beginning of the main handler. The main handler can then set the instance in the context for the access and main handler to use.

MusicDin replied:
Discussion moved to https://github.com/canonical/lxd/pull/16329

MusicDin commented:
Based on @markylaing comments, also fixed:
- Rename constant `devLXDSecurityMgmtVolumesKey` -> `devLXDSecurityManagementVolumesKey`
- Remove unqoting of ETags, they are set in the header as is. The ETags from LXD are always quoted anyway.
- Added several comments.

tomponline commented on lxd/devlxd_response_render.go:
```diff
@@ -21,3 +21,24 @@ func Render(req *http.Request, resp response.Response) error {
 }
+
+// RenderToStruct renders the response into a struct and returns the ETag.
+func RenderToStruct(req *http.Request, resp response.Response, target any) (etag string, err error) {
```
not devlxd specific, but in devlxd file.

tomponline commented on lxd/request/request.go:
```diff
@@ -1,14 +1,74 @@
+	var err error
+
+	// Create new HTTP request.
+	if data != nil {
```
This section looks very similar to that of client.RawQuery.

Can we de-dupe it somehow? Perhaps by moving this to the client package and using it in LXD?

MusicDin replied:
Fixed. Also de-duped from DevLXD and LXD clients.

MusicDin commented on shared/api/devlxd.go:
```diff
@@ -41,6 +41,21 @@ type DevLXDGet struct {
 	Auth string `json:"auth" yaml:"auth"`
+
+	// List of supported storage drivers.
+	// Example: [{"name": "zfs", "remote": false}, {"name": "btrfs", "remote": true}]
```
Any naming suggestion? How about `devlxd_volume_management` to cover new API as well?

tomponline replied:
still missing API extension

MusicDin replied:
Now it is really done :)

tomponline commented on lxd/devlxd_instances.go:
```diff
@@ -3,13 +3,78 @@ package main
+	// Fetch instance.
+	targetInst := api.Instance{}
+
+	url := api.NewURL().Path("1.0", "instances", targetInstName).Project(projectName).WithQuery("recursion", "1").URL
```
No need for `.URL` on the end, you can just use `api.NewURL()` returned and call `.String()`  on it when needed.

tomponline commented on client/devlxd.go:
```diff
@@ -39,6 +39,10 @@ type ProtocolDevLXD struct {
 	bearerToken string
+
+	// clusterTarget is the name of the cluster member
+	// that this client is targeting.
```
no need to overflow the line here

MusicDin replied:
Fixed

tomponline commented on lxd/devlxd_storage.go:
```diff
@@ -337,7 +337,8 @@ func devLXDStoragePoolVolumePutHandler(d *Daemon, r *http.Request) response.Resp
 
 	// Ensure the volume owner cannot be changed.
-	if !isDevLXDVolumeOwner(vol.Config, identity.Identifier) {
+	owner, ok := vol.Config["volatile.devlxd.owner"]
```
why not using helper function here?

tomponline replied:
I prefer the approach you proposed that uses isDevLXDVolumeOwner() so we dont duplicate logic.

tomponline replied:
And we can see all use-case of checking device ownership.

tomponline commented:
> The only open question remaining is:
> 
> > @markylaing @MusicDin why isn't this being done by an API route level access handler like in storagePoolVolumeTypeCmd?
> 
> This is comment on adding volume details to the request in `GET/PUT/DELETE storage-pools/{poolName}/volumes/{type}/{volumeName}` DevLXD handlers:
> 
> ```go
> err = addStoragePoolVolumeDetailsToRequestContext(d.State(), req)
> if err != nil {
>         return response.DevLXDErrorResponse(err)
> }
> ```
> 
> Based on our previous discussion, the storage details are not added during in the main API handlers because the target needs to be resolved beforehand. However, we do not enforce that the target is set in DevLXD either, so this means that target is effectively not resolved prior calling any of those endpoints.
> 
> However, I would expect the (main LXD) handler to error out if multiple storage volumes are found, but the target is not provided. Effectively enforcing target requirement if needed.
> 
> I'm open to clarifications and corrections :)

Lets discuss at team meeting today as i still dont see why devlxd needs to be different to main API handler here.

MusicDin commented:
@tomponline If the above issue is the only one remaining, could we get that merged, and I can address the issue after our discussions?

This would unblock me for the CSI itself, and all subsequent PRs will be much smaller and easier to review.

simondeziel commented on doc/api-extensions.md:
```diff
@@ -2711,5 +2711,22 @@ See {ref}`DevLXD bearer tokens <devlxd-authentication-bearer>`.
+Volumes and instance devices created through the DevLXD are marked as owned by the authenticated identity and can later be modified or removed through DevLXD only by the same identity.
+
+New DevLXD endpoints:
+- `GET /1.0/storage-pools/{pool}` — Retrieve a storage pool.
+- `GET /1.0/storage-pools/{pool}/volumes` — List owned custom volumes in a pool.
+- `GET /1.0/storage-pools/{pool}/volumes/{volume}` — Retrieve an owned custom volume.
+- `POST /1.0/storage-pools/{pool}/volumes` — Create a new owned custom volume in a pool.
+- `PUT /1.0/storage-pools/{pool}/volumes/{volume}` — Update an owned custom volume.
+- `DELETE /1.0/storage-pools/{pool}/volumes/{volume}` — Delete an owned custom volume.
+- `GET /1.0/instances/{inst}` — Retrieve an instance and its owned devices.
+- `PATCH /1.0/instances/{inst}` — Add new instance devices or modify owned devices.
```
Could you please run `make doc-lint` as I think those `- ` prefixes seem to upset it.

Maybe using `* ` would fare better?

MusicDin replied:
Thanks. That was the issue + missing line before list elements.

markylaing commented on lxd/devlxd_storage.go:
```diff
@@ -0,0 +1,513 @@
+var devLXDStoragePoolVolumesTypeEndpoint = APIEndpoint{
+	Path: "storage-pools/{poolName}/volumes/{type}",
+	Get:  APIEndpointAction{Handler: devLXDStoragePoolVolumesGetHandler, AccessHandler: allowDevLXDAuthenticated},
+	Post: APIEndpointAction{Handler: devLXDStoragePoolVolumesPostHandler, AccessHandler: allowDevLXDAuthenticated},
```
This access handler should be checking if the devlxd identity has permission to create volumes in the project.

markylaing commented on lxd/devlxd_storage.go:
```diff
@@ -0,0 +1,513 @@
+var devLXDStoragePoolVolumesEndpoint = APIEndpoint{
+	Path: "storage-pools/{poolName}/volumes",
+	Get:  APIEndpointAction{Handler: devLXDStoragePoolVolumesGetHandler, AccessHandler: allowDevLXDAuthenticated},
+	Post: APIEndpointAction{Handler: devLXDStoragePoolVolumesPostHandler, AccessHandler: allowDevLXDAuthenticated},
```
This access handler should be checking if the devlxd identity has permission to create volumes in the project.

markylaing commented on lxd/devlxd_storage.go:
```diff
@@ -0,0 +1,513 @@
+	return response.DevLXDResponse(http.StatusOK, "", "raw")
+}
+
+var devLXDStoragePoolVolumeTypeEndpoint = APIEndpoint{
```
Can you please move all of these `APIEndpoint` definitions to the top of the file.

markylaing reviewed (changes requested):
Most handlers look fine from an auth perspective except for creation of storage volumes.
Even with the storage volume management security key, we should not allow creation of storage volumes unless the identity has `can_create_storage_volumes` on the project.

markylaing commented on lxd/devlxd_storage.go:
```diff
@@ -0,0 +1,513 @@
+var devLXDStoragePoolVolumesEndpoint = APIEndpoint{
+	Path: "storage-pools/{poolName}/volumes",
+	Get:  APIEndpointAction{Handler: devLXDStoragePoolVolumesGetHandler, AccessHandler: allowDevLXDAuthenticated},
+	Post: APIEndpointAction{Handler: devLXDStoragePoolVolumesPostHandler, AccessHandler: allowDevLXDPermission(entity.TypeProject, auth.EntitlementCanCreateStorageVolumes)},
```
Nice, thanks
